syntax = "proto3";

option go_package = ".;pb";

import "google/protobuf/empty.proto";

message PortForwardingIdentifier {
  string host = 1;
  string protocol = 2;
  uint32 source_port = 3;
}

message PortForwarding {
  string protocol = 2;
  uint32 source_port = 3;
  string target_ip = 4;
  uint32 target_port = 5;
}

message ListPortForwardingsRequest {
  string host = 1;
}

message ListPortForwardingsResponse {
  repeated PortForwarding port_forwardings = 1;
}

message PutPortForwardingRequest {
  PortForwarding port_forwarding = 1;
  string host = 2;
}

service PortForwardingService {
  rpc GetPortForwarding(PortForwardingIdentifier) returns (PortForwarding) {}
  rpc ListPortForwardings(ListPortForwardingsRequest) returns (ListPortForwardingsResponse) {}
  rpc PutPortForwarding(PutPortForwardingRequest) returns (PortForwarding) {}
  rpc DeletePortForwarding(PortForwardingIdentifier) returns (google.protobuf.Empty) {}
}
